### DISCUSSION TITLE
Too much "CLOSE_WAIT" connections

### DISCUSSION THREAD

**Discussion by paulomanrique** [php-amqplib/php-amqplib#1172] [Q&A]
Hello. We are having an issue where our consumers are leaving a lot of connections in a CLOSE_WAIT state, against the RabbitMq port (we are using AWS RabbitMq).

We are currently running 10 workers (via Supervisord), that must be up at all times, and execute tasks that are quick or slower, depending on then.

This is, roughly the code we are using:

```php
<?php

$sslOptions = [
    'verify_peer' => true,
    'ssl_on' => true,
    'ssl_verify' => true,
];

$options = [
    'connection_timeout' => 240,
    'read_write_timeout' => 240,
    'heartbeat' => 40,
    'keepalive' => true,
];

$credentials = $fetchCredentials();
$connection = new AMQPSSLConnection(
    $credentials->rabbit_host,
    $credentials->rabbit_port,
    $credentials->server,
    $credentials->password,
    '/',
    $sslOptions,
    $options
);

$sender = new PCNTLHeartbeatSender($connection);
$sender->register();
$channel = $connection->channel();

$channel->queue_declare(
    $queue,
    false,
    true,
    false,
    false
);

$channel->basic_qos(
    null,
    1,
    null
);

$channel->basic_consume(
    $queue,
    '',
    false,
    false,
    false,
    false,
    [$this, 'callback']
);

while (count($channel->callbacks)) {
    $channel->wait();
}

$channel->close();
$connection->close();
```

These are some examples on how the alerts we are getting looks like:

![image](https://github.com/php-amqplib/php-amqplib/assets/4464977/8607fade-0e59-44e2-b929-2772b47c18f6)

![image](https://github.com/php-amqplib/php-amqplib/assets/4464977/4d46f408-aa1a-4804-a1ae-8323dfff468f)

Any ideas on why this is happening, and what can we do to prevent/remedy this issue?

**Comment by lukebakken** [comment]
Hello, thanks for using RabbitMQ and this library.

Could you please let us know the versions of all software involved?
* RabbitMQ
* Erlang
* Operating System
* PHP
* `php-amqplib`

Other questions -
* Are you using a load balancer?
* Based on your graphs, it looks as though this environment works normally but then experiences the `CLOSE_WAIT` issue from time to time. What is happening during those times? Does workload increase? Does your application crash and restart? What is logged by RabbitMQ at those times?

**Reply by paulomanrique** [reply]
Hey @lukebakken, thanks for the prompt Answer:

The server is running on AWS, so I don't know the operating system it's running. The versions:
- RabbitMQ 3.11.28
- Erlang 25.3.2.8

The client:
- Operating system is Almalinux 8.9
- PHP is 8.2.18
- php-amqplib is 3.6.2

--------------------

- AWS is using a load balancer
- Yes, the workload of messages increase... I don't think it's crashing and restarting, this started to happen once we enabled the heartbeat to keep the connections alive.
- I don't know if we have access to RabbitMq logs on AWS, I will take a look and find out.

**Reply by lukebakken** [reply]
Just so you know, RabbitMQ 3.11 is out of community support.

We need to know what RabbitMQ logs at the time of the CLOSE_WAIT issue.

You could also try _not_ using the load balancer, if possible.

**Reply by paulomanrique** [reply]
I will check if we can update that instance on AWS and also try to not use the load balancer.

Just to clarify: when I talk about the CLOSE_WAIT issue, it's happening on the client side, not on server side. Do you still want logs from the server, when that is happening?

**Reply by lukebakken** [reply]
> Do you still want logs from the server, when that is happening?

Yep! We're looking for log messages like `client closed connection unexpectedly`

**Reply by paulomanrique** [reply]
Okay, I will try to get to that, and also try the things you mentioned and get back to you. 

Thanks again! :)
